pkg/util/objectwatcher/objectwatcher.go

```diff
@@ -97,6 +101,25 @@ func (o *objectWatcherImpl) Create(clusterName string, desireObj *unstructured.U
 	return nil
 }
 
+func (o *objectWatcherImpl) retainClusterFields(desiredObj, clusterObj *unstructured.Unstructured) (*unstructured.Unstructured, error) {
+	// Pass the same ResourceVersion as in the cluster object for update operation, otherwise operation will fail.
+	desiredObj.SetResourceVersion(clusterObj.GetResourceVersion())
+
+	// Retain finalizers since they will typically be set by
+	// controllers in a member cluster.  It is still possible to set the fields
+	// via overrides.
+	desiredObj.SetFinalizers(clusterObj.GetFinalizers())
+	// Merge annotations since they will typically be set by controllers in a member cluster
+	// and be set by user in karmada-controller-plane.
+	util.MergeAnnotations(desiredObj, clusterObj)
+
+	if o.resourceExplorer.HookEnabled(desiredObj, configv1alpha1.ExploreRetaining) {
+		return o.resourceExplorer.Retain(desiredObj, clusterObj)
```

Comment:
`o.resourceExplorer.Retain(desiredObj, clusterObj)` can be assigned to desiredObj when there is no err, and the function `retainClusterFields` can only return error:
```
retainClusterFields(desiredObj, clusterObj *unstructured.Unstructured) error
```